dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,260 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Ticker;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Random;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Routes all requests to one host until an error is received, then we'll move on to the next host.
+ * Upside: ensures clients will always hit warm caches
+ * Downsides:
+ * <ul>
+ *     <li>a single high volume client can result in unbalanced utilization of server nodes.
+ *     <li>when a node is restarted (and clients see 500s/connect exceptions), they will _all_ pin to a different node
+ * </ul>
+ *
+ * To alleviate the second downside, we reshuffle all nodes every 10 minutes.
+ */
+public final class PinUntilErrorChannel implements LimitedChannel {
+    private static final Logger log = LoggerFactory.getLogger(PinUntilErrorChannel.class);
+
+    // we also add some jitter to ensure that there isn't a big spike of reshuffling every 10 minutes.
+    private static final Duration RESHUFFLE_EVERY = Duration.ofMinutes(10);
+
+    // TODO(dfox): could we make this endpoint-specific?
+    private final AtomicInteger currentHost = new AtomicInteger(0); // always positive
+    private final NodeList maybeShufflingNodeList;
+
+    @VisibleForTesting
+    PinUntilErrorChannel(NodeList nodeList) {
+        this.maybeShufflingNodeList = nodeList;
+    }
+
+    public static PinUntilErrorChannel pinUntilError(List<LimitedChannel> channels) {
+        Preconditions.checkArgument(!channels.isEmpty(), "List of channels must not be empty");
+        ReshufflingNodeList shufflingNodeList =
+                new ReshufflingNodeList(channels, ThreadLocalRandom.current(), System::nanoTime);
```

Comment:
bug: ThreadLocalRandom.current() returns the caller threads ThreadLocalRandom, the node list may be used by another thread, potentially causing contention. I wrote a simple `SafeThreadLocalRandom` random at one point that internally delegates to the current threads instance for all calls but never pushed it out broadly.
For now we may want a `Supplier<Random>`